### ISSUE TITLE
Proposal: Helm Developer Workflow

### ISSUE THREAD

**Issue by thomastaylor312** [kubernetes/helm#2069]
# Helm Developer Workflow
This proposal lays out a more detailed structure for how issues are handled and how code is maintained in the Helm repo. Due to the increasing number of users and contributors (yay!), we want to streamline and clarify how we run the project.

As a note: I will keep updating the proposal as feedback comes in.

## Issues
Issues are used as the primary method for tracking anything to do with the Helm project.

## Milestones
For clarity, rename the triage milestones to `Upcoming - Minor` and `Upcoming - Major`

For example, if the current version is `2.2.0` an issue/PR could fall in to one of 4 different active milestones: `2.2.1`, `2.3.0`, `Upcoming - Minor`, or `Upcoming - Major`. If an issue pertains to a specific upcoming bug or minor release, it would go into `2.2.1` or `2.3.0`. If the issue/PR does not have a specific milestone yet, but it is likely that it will land in a `2.X` release, it should go into `Upcoming - Minor`. If the issue/PR is a large functionality add or change and/or it breaks compatibility, then it should be added to the `Upcoming - Major` milestone. An issue that we are not sure we will be doing will not be added to any milestone.

A milestone (and hence release) is considered done when all outstanding issues/PRs have been closed or moved to another milestone.

### Issue Types
There are 4 types of issues (each with their own corresponding label):
- Question: These are support or functionality inquiries that we want to have a record of for future reference. Generally these are questions that are too complex or large to store in the Slack channel or have particular interest to the community as a whole. Depending on the discussion, these can turn into "Feature" or "Bug" issues
- Proposal: Used for items (like this one) that propose a new ideas or functionality that require a larger community discussion. This allows for feedback from others in the community before a feature is actually  developed. This is not needed for small additions. Final word on whether or not a feature needs a proposal is up to the core maintainers. All issues that are proposals should both have a label and an issue title of "Proposal: [the rest of the title]." A proposal can become a "Feature." and do not require a milestone.
- Features: These track specific feature requests and ideas until they are complete. They can evolve from a "Proposal" or can be submitted individually depending on the size.
- Bugs: These track bugs with the code or problems with the documentation (i.e. missing or incomplete)

### Issue Lifecycle
The issue lifecycle is mainly driven by the core maintainers, but is good information for those contributing to the project. All issue types follow the same general lifecycle. Differences are noted below.
1. Issue creation
2. Triage
    - The maintainer in charge of triaging will apply the proper labels for the issue. This includes labels for priority, type, and metadata (such as "starter"). The only issue priority we will be tracking is whether or not the issue is "critical." If additional levels are needed in the future, we can add them.
    - (If needed) Clean up the title to succinctly and clearly state the issue. Also ensure that proposals are prefaced with "Proposal".
    - Add the issue to the correct milestone. If any questions come up, don't worry about adding the issue to a milestone until the questions are answered.
    - This process should be done at least once per work day.
3. Discussion
    - "Feature" and "Bug" issues should eventually be connected to the PR that resolves it. This is the current pattern we follow.
    - Whoever is working on a "Feature" or "Bug" issue (whether a maintainer or someone from the community), should either assign the issue to them self or make a comment in the issue saying that they are taking it.
    - "Proposal" and "Question" issues should stay open until resolved or if they have not been active for more than 30 days. This will help keep the issue queue to a manageable size and reduce noise. Should the issue need to stay open, the `keep open` label can be added.
4. Issue closure

## Pull Requests
Like any good open source project, we use Pull Requests to track code changes

### PR Lifecycle
1. PR creation
    - We more than welcome PRs that are currently in progress. They are a great way to keep track of important work that is in-flight, but useful for others to see. If a PR is a work in progress, it **must** be prefaced with "WIP: [title]". Once the PR is ready for review, remove "WIP" from the title.
    - It is preferred, but not required, to have a PR tied to a specific issue.
2. Triage
    - The maintainer in charge of triaging will apply the proper labels for the issue. This should include at least a size label, `bug` or `feature`, and `awaiting review` once all labels are applied. See the "Labels" section for full details on the definitions of labels
    - Add the PR to the correct milestone. Generally, this should be the same as the issue the PR closes.
3. Assigning reviews
    - Once a review has the `awaiting review` label, maintainers will review them as schedule permits. Whoever takes the issue should self-request a review. If this method does not work, further discussion will be required to figure out the best method for assigning reviews.
    - Reviews from others in the community, especially those who have encountered a bug or have requested a feature, are highly encouraged, but not required. Maintainer reviews **are** required before any merge
    - Any PR with the `size/large` label requires 2 review approvals from maintainers before it can be merged. Those with `size/medium` are per the judgement of the maintainers
4. Reviewing/Discussion
    - Once a maintainer begins reviewing a PR, they should remove the `awaiting review` label and add the `in progress` label so the person submitting knows that it is being worked on. This is especially helpful when the review may take awhile.
    - All reviews should be completed using Github review tool. Our current pattern of a simple "LGTM" will be deprecated in favor of "Approved" reviews.
    - A "Comment" review should be used when there are questions about the code that should be answered, but that don't involve code changes. This type of review does not count as approval.
    - A "Changes Requested" review indicates that changes to the code need to be made before they will be merged.
    - Reviewers should update labels as needed (such as `needs rebase`)
5. Address comments by answering questions or changing code
6. Merge or close
    - PRs should stay open until merged or if they have not been active for more than 30 days. This will help keep the PR queue to a manageable size and reduce noise. Should the PR need to stay open (like in the case of a WIP), the `keep open` label can be added.

#### Documentation PRs
Documentation PRs will follow the same lifecycle as other PRs. They will also be labeled with the `docs` label. For documentation, special attention will be paid to spelling, grammar, and clarity (whereas those things don't matter *as* much for comments in code).

## Maintainer Responsibilities
The new workflow will add one additional responsibility for core maintainers. Each week, one person will serve as the designated "triager" starting after the public standup meeting on Thursday. This person will be in charge triaging new PRs and issues thoughout the work week. Should this prove to be to onerous, we can adjust as needed.

## Labels
The following tables define all label types and will be added to the developer documentation upon approval of the proposal

### Common
| Label | Description |
| ----- | ----------- |
| `bug` | Marks an issue as a bug or a PR as a bugfix |
| `critical` | Marks an issue or PR as critical. This means that addressing the PR or issue is top priority and will be handled first by maintainers |
| `docs` | Indicates the issue or PR is a documentation change |
| `duplicate` | Indicates that the issue or PR is a duplicate of another |
| `feature` | Marks the issue as a feature request or a PR as a feature implementation |
| `keep open` | Denotes that the issue or PR should be kept open past 30 days of inactivity |
| `refactor` | Indicates that the issue is a code refactor and is not fixing a bug or adding additional functionality |

### Issue Specific
| Label | Description |
| ----- | ----------- |
| `help wanted` | This issue is one the core maintainers cannot get to right now and would appreciate help with |
| `proposal` | This issue is a proposal |
| `question/support` | This issue is a support request or question |
| `starter` | This issue is a good for someone new to contributing to Helm |
| `wont fix` | The issue has been discussed and will not be implemented (or accepted in the case of a proposal) |

### PR Specific
| Label | Description |
| ----- | ----------- |
| `awaiting review` | The PR has been triaged and is ready for someone to review |
| `breaking` | The PR has breaking changes (such as API changes) |
| `cncf-cla: no` | The PR submitter has **not** signed the project CLA. |
| `cncf-cla: yes` | The PR submitter has signed the project CLA. This is required to merge. |
| `in progress` | Indicates that a maintainer is looking at the PR, even if no review has been posted yet |
| `needs pick` | Indicates that the PR needs to be picked into a feature branch (generally bugfix branches). Once it has been, the `picked` label should be applied and this one removed |
| `needs rebase` | A helper label used to indicate that the PR needs to be rebased before it can be merged. Used for easy filtering |
| `picked` | This PR has been picked into a feature branch |

#### Size labels
Size labels are used to indicate how "dangerous" a PR is. The guidelines below are used to assign the labels, but ultimately this can be changed by the maintainers. For example, even if a PR only makes 30 lines of changes in 1 file, but it changes key functionality, it will likely be labeled as `size/large` because it requires sign off from multiple people. Conversely, a PR that adds a small feature, but requires another 75 lines of tests to cover all cases, could be labeled as `size/small` even though the number lines is greater than defined below.

| Label | Description |
| ----- | ----------- |
| `size/small` | Anything less than or equal to 4 files and 150 lines. Only small amounts of manual testing may be required |
| `size/medium` | Anything greater than `size/small` and less than or equal to 8 files and 300 lines. Manual validation should be required. |
| `size/large` | Anything greater than `size/medium`. This should be thoroughly tested before merging and always requires 2 approvals. This also should be applied to anything that is a significant logic change. |

## Future Ideas
- Label and issue automation. This would automatically take care of labeling some things and automate closing of stale PRs and issues
- Use Github projects to keep track of in progress work and make the whole dev lifecycle more visible to the rest of the community

## Proposal Complete
This proposal will be considered complete when the following items are 

- [x] All core maintainers have given an "LGTM"
- [x] All labels have been created
- [x] All issues/PRs have been updated with new labels or closed based on new workflow
- [x] Developer documentation has been updated with relevant information from this proposal

**Comment by technosophos** [comment]
- I like `upcoming-minor` and `-major` a lot.
- I'm good with setting for issue types, though maybe `help` would be better than `question`... up to you.
- In issue lifecycle: What do we do if we get a feature request, and we're not sure whether we want to do it at all? Do we put it in `upcoming-minor` and punt that decision down the road? Or leave it as is (no tag)? Or do we add a label?
- And can we change "once a day" to "once per workday"?
- We encourage community code reviews, but don't specify whether they count toward the final Approved count of one or two LGTMs.
- I like `picked` and `needs pick` much more than the current labels.

I think we should remain pretty consistent that one or two _maintainers_ need to approve a PR. But that community code reviews are definitely brought into consideration.

Overall, I _love_ this proposal and think it would greatly reduce the churn we're feeling right now. So I'm happy to weigh in with my LGTM now.

**Comment by thomastaylor312** [comment]
@technosophos Thanks for all of the feedback! I added all of the changes you requested including clarifying that we monitor the queue on work days and not every day (phew...) 😄 . Also, I think issues that we aren't sure if we want in but are interested in will not have a milestone (which I added in the proposal) and we can use the `keep open` label to keep them open and in view

**Comment by technosophos** [comment]
@thomastaylor312 One other thing that I was thinking about this morning was whether we should use the term `backlog` instead of `upcoming`.

Pros:
- similar to Agile terminology
- doesn't imply that we're doing it "right away"

Cons:
- "backlog-minor" and "backlog-major" is sort of verbose
- "backlog" might sound like we're just delaying

**Comment by thomastaylor312** [comment]
I am ok with either, though I prefer `upcoming` personally.

**Comment by thomastaylor312** [comment]
Per feedback from @michelleN in the public dev standup, I added a `refactor` label for issues and PRs

**Comment by bacongobbler** [comment]
Hey! Just wanted to say this is great work. I got a couple comments about the new milestone changes, mostly for clarification:

> For clarity, rename the triage milestones to `Upcoming - Minor` and `Upcoming - Major`

Would these milestones become a rolling target? i.e. they have a due date, but perpetually exist.

There comes a small window usually near the end of a release in which a developer would like to triage issues into the *next* minor/major release, but still need to maintain the current milestone for the release coming out. For example, if a feature is being punted to the next release, would we de-milestone the issue, then re-add it back to `Upcoming - Minor` afterwards?

Just for clarity, is `minor` meant to indicate a minor release and `major` to indicate a major release? Are patch releases not tracked in milestones or are they tracked ad-hoc? I feel like having some process of creating an `Upcoming - Patch` milestone may be useful in some cases.

**Comment by bacongobbler** [comment]
Also it may not apply to y'all but we @ deis thought it may be useful to write a tool to codify creating and changing labels/milestones for the project. Whether or not it's useful for you is completely up to you, but I thought I'd throw it out there :)

https://github.com/deis/seed-repo

**Comment by adamreese** [comment]
I would like to cast my LGTM.

**Comment by thomastaylor312** [comment]
@bacongobbler Thanks for the feedback. So to answer your questions, I think all patch PRs will go into the latest patch release milestone (which should always exist). Because if they are a patch, they shouldn't be punted.

As for the issues at the end of the release cycle, basically how I see it working is that if it goes into the current minor release it goes into the (for example) `2.3.0` release and one that would go into `2.4.0` would be put in that milestone if it is created at `Upcoming - Minor` if it is not. That make any more sense?

And the `Upcoming` milestones would be rolling targets like you said. Do you have a better idea of how to manage those?

**Comment by bacongobbler** [comment]
> That make any more sense?

Crystal clear.

> Do you have a better idea of how to manage those?

Nope! I don't have a better idea on how to manage them. We've been happy opening `vX.Y.Z` milestones and just closing them when we're done, then open a `vX.Y+1.Z` milestone for the next release after vX.Y.Z is out. Just personally curious on your team's experience and the backstory that is prompting the move from a "deadline target" milestone to a "rolling target" milestone is all. :)

**Comment by bacongobbler** [comment]
This proposal also has my unofficial LGTM as well, just wanting to learn from your experiences :)

**Comment by scottrigby** [comment]
My 2¢ is this looks great so far 👍  I do like that `Question` issue types can turn into other types depending on the discussion. 

Because questions often help identify gaps (ambiguity, lack of examples, etc) in documentation, should the workflow be that an issue starts out as a `Question` and then if documentation needs updating it then turns into a `Bug`? Not sure `Bug` is a good description for the need to document (or clarify docs). Would it be helpful to have a general `Task` issue type? I know that has been helpful for the Drupal community's issue queue.

**Comment by thomastaylor312** [comment]
@scottrigby Good point. I think it would end up labeling it with `docs` if it was a documentation update.

**Comment by beeradb** [comment]
The timing of this is quite apropos as internally my organization has been working towards documenting our workflow and it looks very very similar. 

To respond to @technosophos:

> In issue lifecycle: What do we do if we get a feature request, and we're not sure whether we want to do it at all? Do we put it in upcoming-minor and punt that decision down the road? Or leave it as is (no tag)? Or do we add a label?

Internally we group issues into three mutually exclusive buckets. Those are:

<dl>
<dt>Hibernate</dt>
<dd>An issue which may have merit and we want to track, but we're unlikely to work on soon. We periodically review the hibernate list and will move them to incubate if we feel they are ready to be worked on.</dd>
<dt>Incubate</dt>
<dd>An issue that we plan on working on soon (i.e. upcoming-major or upcoming-minor) which still needs improved planning or done criteria. Any issue in this state is actively being planned and will make it into an upcoming sprint</dd>
<dt>Actionable</dt>
<dd>A defined issue which is ready to be worked on. For us, only actionable tickets are added to sprints</dd>
</dl>

Obviously, things might change a bit when working towards releases rather than sprints, but just thought I'd share how we deal with that issue, as we track pretty closely with this proposal.

**Comment by thomastaylor312** [comment]
@beeradb I like this a lot, but I think it will probably become more useful if we transition to using Github Projects down the road. Really good idea though

**Comment by michelleN** [comment]
LGTM from me. looks good! thanks!

**Comment by prydonius** [comment]
LGTM, the PR review process is pretty similar to what we do over in https://github.com/kubernetes/charts. Once the documentation is updated here, I'll take this as a base to create similar documentation in the charts repo.

Thanks a lot, @thomastaylor312!

**Comment by thomastaylor312** [comment]
Looks like we are good from the LGTM front. I'll begin implementing the other steps and track them in the main proposal above

**Comment by thomastaylor312** [comment]
I have the labels created and PRs all sorted out. I'll be working on the issue queue tomorrow and over the weekend
